[quantization] quantized torch.topk

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26486 [quantization] quantized torch.topk**

This PR adds a quantized version of torch.topk, supporting all the same options

Benchmark script
```
import torch
import time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    X = torch.rand(6, 5, 1024)
    qX = torch.quantize_linear(X, 0.01, 0, dtype)
    X = qX.dequantize()
    
    NITER = 10000

    s = time.time()
    for i in range(NITER):
        float_out = torch.topk(X, 50)
    float_time_per_iter = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.topk(qX, 50)
    quant_time_per_iter = (time.time() - s) / NITER

    print(dtype)
    print('float ms', 'quant ms', 'float gB/s', 'quant gB/s', sep='\t')
    nbytes_float = (X.numel() + float_out[0].numel()) * X.element_size()
    nbytes_quant = (qX.numel() + quant_out[0].numel()) * qX.element_size()
    print(float_time_per_iter * 1000,
          quant_time_per_iter * 1000,
          nbytes_float / float_time_per_iter / 1e9,
          nbytes_quant / quant_time_per_iter / 1e9, sep='\t')
```

Results

```
torch.qint8
float ms	quant ms	float gB/s	quant gB/s
0.3706729888916016	0.3370296716690064	0.34769191136743244	0.09559989136992947
torch.quint8
float ms	quant ms	float gB/s	quant gB/s
0.38260042667388916	0.34079675674438475	0.3368527346412275	0.09454315325003715
torch.qint32
float ms	quant ms	float gB/s	quant gB/s
0.38033516407012935	0.3364055633544922	0.3388590174539739	0.38310900305828427

```

Differential Revision: [D17529988](https://our.internmc.facebook.com/intern/diff/D17529988)

Commit:
[quantization] quantized torch.topk

Commit:
Update on "[quantization] quantized torch.topk"


This PR adds a quantized version of torch.topk, supporting all the same options

Benchmark script
```
import torch
import time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    X = torch.rand(6, 5, 1024)
    qX = torch.quantize_linear(X, 0.01, 0, dtype)
    X = qX.dequantize()
    
    NITER = 10000

    s = time.time()
    for i in range(NITER):
        float_out = torch.topk(X, 50)
    float_time_per_iter = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.topk(qX, 50)
    quant_time_per_iter = (time.time() - s) / NITER

    print(dtype)
    print('float ms', 'quant ms', 'float gB/s', 'quant gB/s', sep='\t')
    nbytes_float = (X.numel() + float_out[0].numel()) * X.element_size()
    nbytes_quant = (qX.numel() + quant_out[0].numel()) * qX.element_size()
    print(float_time_per_iter * 1000,
          quant_time_per_iter * 1000,
          nbytes_float / float_time_per_iter / 1e9,
          nbytes_quant / quant_time_per_iter / 1e9, sep='\t')
```

Results

```
torch.qint8
float ms	quant ms	float gB/s	quant gB/s
0.3706729888916016	0.3370296716690064	0.34769191136743244	0.09559989136992947
torch.quint8
float ms	quant ms	float gB/s	quant gB/s
0.38260042667388916	0.34079675674438475	0.3368527346412275	0.09454315325003715
torch.qint32
float ms	quant ms	float gB/s	quant gB/s
0.38033516407012935	0.3364055633544922	0.3388590174539739	0.38310900305828427

```

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -218,7 +219,80 @@ void qmaxpool_2d_nhwc_kernel(const Tensor &qx,
-} 
+}
+
+void qtopk_kernel(Tensor& values,
```
can't you literally use the existing function and define _isnan<quint8> to be always false? but I'm fine with duplication as is as well

jamesr66a replied:
The current macros we have for dispatch are `AT_DISPATCH_ALL_TYPES` and `AT_DISPATCH_QINT_TYPES`. The former does not include qint types and I didn't want to add yet another dispatch macro so I just copied the code here

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qsort.cpp:
```diff
@@ -0,0 +1,51 @@
+    int64_t dim,
+    bool largest,
+    bool sorted) {
+  Tensor values = at::_empty_affine_quantized(
```
can you call empty like maybe?

jamesr66a replied:
We don't want to allocate the same size as input, though

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qsort.cpp:
```diff
@@ -0,0 +1,51 @@
+namespace at {
+namespace native {
+
+std::tuple<Tensor&, Tensor&> quantized_topk_out_cpu(
```
what are the semantics of _out version? do we generally use it's quantizer? We need to either enforce quantizer to be the same or reset quantize in the output. cc @jerryzh168  on what we usually support in other ops

jamesr66a replied:
Hmm, good question. Exposing the _out variant implies doing a requantization in the kernel. I think I can just not expose the _out variant as an op and only use it internally, thus ensuring the quantizer for the input and output are the same

jerryzh168 replied:
`_out` variant is a way of passing output `scale` and `zero_point` right now.
the initial motivation is to reuse the API we already have in torch namespace, since many torch functions have an out variant. 
e.g.
```
add(a, b, out_scale, out_zp)
```
```
out = empty_quantized(out_scale, out_zp)
add_out(a, b, out)
```

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qsort.cpp:
```diff
@@ -0,0 +1,51 @@
+  Tensor values = at::_empty_affine_quantized(
+    {0},
+    self.options(),
+    self.q_scale(),
```
this will assert that quantizer is uniform, but it might be nice to add an explicit TORCH_CHECK for it

dzhulgakov commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -982,6 +982,7 @@
   dispatch:
     CPU: resize_cpu_
     CUDA: resize_cuda_
+    QuantizedCPU: resize_cpu_
```
hm, can you check that it throws if the input quantizer is per-channel? It's not safe to invoke it in this case

jamesr66a replied:
I think I can but it probably involves making a specific quantized resize function rather than reusing this one

llyfacebook commented:
By the way, have you thought about implementing it with priority queue? Given the k is not that large in practice.

Commit:
Update on "[quantization] quantized torch.topk"


This PR adds a quantized version of torch.topk, supporting all the same options

Benchmark script
```
import torch
import time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    X = torch.rand(6, 5, 1024)
    qX = torch.quantize_linear(X, 0.01, 0, dtype)
    X = qX.dequantize()
    
    NITER = 10000

    s = time.time()
    for i in range(NITER):
        float_out = torch.topk(X, 50)
    float_time_per_iter = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.topk(qX, 50)
    quant_time_per_iter = (time.time() - s) / NITER

    print(dtype)
    print('float ms', 'quant ms', 'float gB/s', 'quant gB/s', sep='\t')
    nbytes_float = (X.numel() + float_out[0].numel()) * X.element_size()
    nbytes_quant = (qX.numel() + quant_out[0].numel()) * qX.element_size()
    print(float_time_per_iter * 1000,
          quant_time_per_iter * 1000,
          nbytes_float / float_time_per_iter / 1e9,
          nbytes_quant / quant_time_per_iter / 1e9, sep='\t')
```

Results

```
torch.qint8
float ms	quant ms	float gB/s	quant gB/s
0.3706729888916016	0.3370296716690064	0.34769191136743244	0.09559989136992947
torch.quint8
float ms	quant ms	float gB/s	quant gB/s
0.38260042667388916	0.34079675674438475	0.3368527346412275	0.09454315325003715
torch.qint32
float ms	quant ms	float gB/s	quant gB/s
0.38033516407012935	0.3364055633544922	0.3388590174539739	0.38310900305828427

```

Commit:
Update on "[quantization] quantized torch.topk"


This PR adds a quantized version of torch.topk, supporting all the same options

Benchmark script
```
import torch
import time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    X = torch.rand(6, 5, 1024)
    qX = torch.quantize_linear(X, 0.01, 0, dtype)
    X = qX.dequantize()
    
    NITER = 10000

    s = time.time()
    for i in range(NITER):
        float_out = torch.topk(X, 50)
    float_time_per_iter = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.topk(qX, 50)
    quant_time_per_iter = (time.time() - s) / NITER

    print(dtype)
    print('float ms', 'quant ms', 'float gB/s', 'quant gB/s', sep='\t')
    nbytes_float = (X.numel() + float_out[0].numel()) * X.element_size()
    nbytes_quant = (qX.numel() + quant_out[0].numel()) * qX.element_size()
    print(float_time_per_iter * 1000,
          quant_time_per_iter * 1000,
          nbytes_float / float_time_per_iter / 1e9,
          nbytes_quant / quant_time_per_iter / 1e9, sep='\t')
```

Results

```
torch.qint8
float ms	quant ms	float gB/s	quant gB/s
0.3706729888916016	0.3370296716690064	0.34769191136743244	0.09559989136992947
torch.quint8
float ms	quant ms	float gB/s	quant gB/s
0.38260042667388916	0.34079675674438475	0.3368527346412275	0.09454315325003715
torch.qint32
float ms	quant ms	float gB/s	quant gB/s
0.38033516407012935	0.3364055633544922	0.3388590174539739	0.38310900305828427

```

Commit:
Update on "[quantization] quantized torch.topk"


This PR adds a quantized version of torch.topk, supporting all the same options

Benchmark script
```
import torch
import time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    X = torch.rand(6, 5, 1024)
    qX = torch.quantize_linear(X, 0.01, 0, dtype)
    X = qX.dequantize()
    
    NITER = 10000

    s = time.time()
    for i in range(NITER):
        float_out = torch.topk(X, 50)
    float_time_per_iter = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.topk(qX, 50)
    quant_time_per_iter = (time.time() - s) / NITER

    print(dtype)
    print('float ms', 'quant ms', 'float gB/s', 'quant gB/s', sep='\t')
    nbytes_float = (X.numel() + float_out[0].numel()) * X.element_size()
    nbytes_quant = (qX.numel() + quant_out[0].numel()) * qX.element_size()
    print(float_time_per_iter * 1000,
          quant_time_per_iter * 1000,
          nbytes_float / float_time_per_iter / 1e9,
          nbytes_quant / quant_time_per_iter / 1e9, sep='\t')
```

Results

```
torch.qint8
float ms	quant ms	float gB/s	quant gB/s
0.3706729888916016	0.3370296716690064	0.34769191136743244	0.09559989136992947
torch.quint8
float ms	quant ms	float gB/s	quant gB/s
0.38260042667388916	0.34079675674438475	0.3368527346412275	0.09454315325003715
torch.qint32
float ms	quant ms	float gB/s	quant gB/s
0.38033516407012935	0.3364055633544922	0.3388590174539739	0.38310900305828427

```

lly-zero-one reviewed (commented):
Is there speedup for topk with AVX2 flag? If not, we can just leave it under cpu directory?

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -680,6 +680,32 @@ def test_adaptive_avg_pool2d_nhwc(self, X, output_size_h, output_size_w):
+           dim=st.integers(1, 4),
+           largest=st.booleans(),
+           sorted=st.booleans())
+    def test_qtopk(self, X, k, dim, largest, sorted):
```
Can you also add a test to check if this works for NHWC layout?

raghuramank100 reviewed (approved):
Looks good, had a question on verifying functionality for NHWC layout.

Commit:
Update on "[quantization] quantized torch.topk"


This PR adds a quantized version of torch.topk, supporting all the same options

Benchmark script
```
import torch
import time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    X = torch.rand(6, 5, 1024)
    qX = torch.quantize_linear(X, 0.01, 0, dtype)
    X = qX.dequantize()
    
    NITER = 10000

    s = time.time()
    for i in range(NITER):
        float_out = torch.topk(X, 50)
    float_time_per_iter = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.topk(qX, 50)
    quant_time_per_iter = (time.time() - s) / NITER

    print(dtype)
    print('float ms', 'quant ms', 'float gB/s', 'quant gB/s', sep='\t')
    nbytes_float = (X.numel() + float_out[0].numel()) * X.element_size()
    nbytes_quant = (qX.numel() + quant_out[0].numel()) * qX.element_size()
    print(float_time_per_iter * 1000,
          quant_time_per_iter * 1000,
          nbytes_float / float_time_per_iter / 1e9,
          nbytes_quant / quant_time_per_iter / 1e9, sep='\t')
```

Results

```
torch.qint8
float ms	quant ms	float gB/s	quant gB/s
0.3706729888916016	0.3370296716690064	0.34769191136743244	0.09559989136992947
torch.quint8
float ms	quant ms	float gB/s	quant gB/s
0.38260042667388916	0.34079675674438475	0.3368527346412275	0.09454315325003715
torch.qint32
float ms	quant ms	float gB/s	quant gB/s
0.38033516407012935	0.3364055633544922	0.3388590174539739	0.38310900305828427

```

Differential Revision: [D17529988](https://our.internmc.facebook.com/intern/diff/D17529988)

Commit:
Update on "[quantization] quantized torch.topk"


This PR adds a quantized version of torch.topk, supporting all the same options

Benchmark script
```
import torch
import time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    X = torch.rand(6, 5, 1024)
    qX = torch.quantize_linear(X, 0.01, 0, dtype)
    X = qX.dequantize()
    
    NITER = 10000

    s = time.time()
    for i in range(NITER):
        float_out = torch.topk(X, 50)
    float_time_per_iter = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.topk(qX, 50)
    quant_time_per_iter = (time.time() - s) / NITER

    print(dtype)
    print('float ms', 'quant ms', 'float gB/s', 'quant gB/s', sep='\t')
    nbytes_float = (X.numel() + float_out[0].numel()) * X.element_size()
    nbytes_quant = (qX.numel() + quant_out[0].numel()) * qX.element_size()
    print(float_time_per_iter * 1000,
          quant_time_per_iter * 1000,
          nbytes_float / float_time_per_iter / 1e9,
          nbytes_quant / quant_time_per_iter / 1e9, sep='\t')
```

Results

```
torch.qint8
float ms	quant ms	float gB/s	quant gB/s
0.3706729888916016	0.3370296716690064	0.34769191136743244	0.09559989136992947
torch.quint8
float ms	quant ms	float gB/s	quant gB/s
0.38260042667388916	0.34079675674438475	0.3368527346412275	0.09454315325003715
torch.qint32
float ms	quant ms	float gB/s	quant gB/s
0.38033516407012935	0.3364055633544922	0.3388590174539739	0.38310900305828427

```

Differential Revision: [D17529988](https://our.internmc.facebook.com/intern/diff/D17529988)
